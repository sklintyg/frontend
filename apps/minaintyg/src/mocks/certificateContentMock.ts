export const certificateContentMock = [
  {
    heading: 'Smittbärarpenning',
    body: `
      <h3>Avstängning enligt smittskyddslagen på grund av smitta</h3>
      <p>Ej angivet</p>
    `,
  },
  {
    heading: 'Grund för medicinskt underlag',
    body: `
      <h3>Intyget är baserat på</h3>
      <h4>Min undersökning av patienten</h4>
      <p>2023-02-07</p>
      <h4>Min telefonkontakt med patienten</h4>
      <p>Ej angivet</p>
      <h4>Journaluppgifter från den</h4>
      <p>Ej angivet</p>
      <h4>Ange vad annat är</h4>
      <p>Ej angivet</p>
    `,
  },
  {
    heading: 'Sysselsättning',
    body: `
      <h3>I relation till vilken sysselsättning bedömer du arbetsförmågan?</h3>
      <ul>
        <li>Nuvarande arbete</li>
      </ul>
      <h3>Ange yrke och arbetsuppgifter</h3>
      <p>Kontorsarbetare</p>

      <table>
        <thead>
          <tr>
            <th>Diagnoskod enligt ICD-10 SE</th>
            <th>Diagnos</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>S829</td>
            <td>Underbensfraktur med ospecificerad lokalisation</td>
          </tr>
        </tbody>
      </table>
    `,
  },
  {
    heading: 'Sjukdomens konsekvenser',
    body: `
      <h3>Funktionsnedsättning</h3>
      <h4>Beskriv undersökningsfynd, testresultat och observationer</h4>
      <p>Patienten har en underbensfraktur. Behandlas med helbensgrips och har därför nedsatt rörlighet i benet. Patienten
        får inte stödja på det skadade benet. Går långsamt och ostadigt med kryckor. Påtalar smärta vid undersökningen.
      </p>
      <h3>Aktivitetsbegränsning</h3>
      <h4>Beskriv vad patienten inte kan göra på grund av sin sjukdom. Ange vad uppgiften grundas
        på.</h4>
      <p>Patienten kan inte gå utan hjälpmedel. Klarar inte av att köra bil. Således mycket svårt att ta sig till arbetet.
        Dessutom problem med trappor på arbetsplatsen. Patienten kan även ha svårt att fokusera på sina arbetsuppgifter
        pga smärta.</p>
    `,
  },
  {
    heading: 'Medicinsk behandling',
    body: `
      <h3>Pågående medicinska behandlingar/åtgärder</h3>
      <h4>Ange vad syftet är och om möjligt tidplan samt ansvarig vårdenhet</h4>
      <p>Erhållit helbensgips, 0-belastning med kryckor. Smärtlindring.</p>
      <h3>Planerade medicinska behandlingar/åtgärder</h3>
      <h4>Ange vad syftet är och om möjligt tidplan samt ansvarig vårdenhet</h4>
      <p>Gips totalt 4 månader, initialt helbensgips. Rtg-kontroll efter 2, 4 och 6 veckor av frakturens läge. Om stabilt
        frakturläge vid 6 v kontroll byte till underbensgips. Rehabilitering efter avslutad gipsbehandling.</p>
    `,
  },
  {
    heading: 'Bedömning',
    body: `
      <h3>Min bedömning av patientens nedsättning av arbetsförmågan</h3>
      <table>
        <thead>
          <tr>
            <th>Nedsättningsgrad</th>
            <th>Från och med</th>
            <th>Till och med</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>100 procent</td>
            <td>2023-02-07</td>
            <td>2023-02-20</td>
          </tr>
        </tbody>
      </table>

      <h3>Patientens arbetsförmåga bedöms nedsatt längre tid än den som Socialstyrelsens
        försäkringsmedicinska beslutsstöd
        anger, därför att</h3>
      <p>Ej angivet</p>

      <h3>Kommer möjligheterna till återgång i arbete försämras om arbetstiden förläggs på annat
        sätt än att arbetstiden
        minskas lika mycket varje dag?</h3>
      <p>Ej angivet</p>

      <h3>Beskriv de medicinska skälen till att möjligheterna till återgång i arbete försämras</h3>
      <p>Ej angivet</p>

      <h3>Resor till och från arbetet med annat färdmedel än normalt kan göra det möjligt för
        patienten att återgå till
        arbetet under sjukskrivningsperioden.</h3>
      <p>Ej angivet</p>

      <h3>Prognos för arbetsförmåga utifrån aktuellt undersökningstillfälle</h3>
      <p>Patienten förväntas kunna återgå helt i nuvarande sysselsättning inom 1 månad.</p>
    `,
  },
  {
    heading: 'Åtgärder',
    body: `
      <h3>Här kan du ange åtgärder som du tror skulle göra det lättare för patienten att återgå i arbete</h3>
      <ul>
        <li>Arbetsanpassning</li>
      </ul>
      <h3>Här kan du ange fler åtgärder. Du kan också beskriva hur åtgärderna kan underlätta återgång i arbete.</h3>
      <p>Möjlighet till distansarbete i hemmet skulle underlätta mycket för patienten.</p>
    `,
  },
  {
    heading: 'Övriga upplysningar',
    body: `
      <h3>Övriga upplysningar</h3>
      <p>Ej angivet</p>
    `,
  },
  {
    heading: 'Kontakt',
    body: `
      <h3>Kontakt med Försäkringskassan</h3>
      <h4>Jag önskar att Försäkringskassan kontaktar mig</h4>
      <p>Ej angivet</p>
      <h4>Ange gärna varför du vill ha kontakt</h4>
      <p>Ej angivet</p>
    `,
  },
].map(({ heading, body }) => ({ heading, body: body.trim().replace(/\s+/g, ' ') }))
