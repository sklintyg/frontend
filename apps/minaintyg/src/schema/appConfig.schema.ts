import { z } from 'zod'

export const appConfigLinkSchema = z.object({
  key: z.string(),
  url: z.string(),
  text: z.string(),
  target: z.string(),
})

export const appConfigSchema = z.object({
  useMinifiedJavascript: z.boolean(),
  version: z.string(),
  buildNumber: z.string(),
  elva77MainUrl: z.string(),
  elva77LoginUrl: z.string(),
  applicationLogoutUrl: z.string(),
  miUserSurveyUrl: z.string(),
  miUserSurveyVersion: z.string(),
  miUserSurveyDateFrom: z.string(),
  miUserSurveyDateTo: z.string(),
  knownRecipients: z.array(z.object({ id: z.string(), name: z.string(), trusted: z.boolean() })),
  links: z.record(appConfigLinkSchema),
  banners: z.array(z.unknown()),
})

export type AppConfig = z.infer<typeof appConfigSchema>
export type AppConfigLink = z.infer<typeof appConfigLinkSchema>
