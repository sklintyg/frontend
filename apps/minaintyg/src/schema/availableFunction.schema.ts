import { z } from 'zod'

export const AvailableFunctionType = z.enum(['SEND_CERTIFICATE'])
export const AvailableFunctionInformationType = z.enum([''])

export const availableFunctionInformationSchema = z.object({
  type: AvailableFunctionInformationType,
  id: z.string(),
})

export const availableFunctionSchema = z.object({
  body: z.string().nullable(),
  description: z.string().nullable(),
  name: z.string().nullable(),
  title: z.string().nullable(),
  type: AvailableFunctionType,
  information: z.array(availableFunctionInformationSchema),
})

export type AvailableFunction = z.infer<typeof availableFunctionSchema>
