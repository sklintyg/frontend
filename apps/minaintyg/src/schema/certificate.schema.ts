import { z } from 'zod'

export const CertificateStatusEnum = z.enum(['NEW', 'REPLACED', 'SENT', 'NOT_SENT'])

export const certificateEventSchema = z.object({
  timestamp: z.string().datetime(),
  certificateId: z.string(),
  description: z.string(),
})

export const certificateSummarySchema = z.object({ label: z.string(), value: z.string() }).optional()

export const certificateTypeSchema = z.object({
  id: z.string(),
  version: z.string(),
  description: z.string().optional(),
})

export const certificateIssuerSchema = z.object({
  name: z.string(),
  phoneNumber: z.string(),
})

export const certificateUnitSchema = z.object({
  id: z.string(),
  name: z.string(),
  address: z.string(),
})

export const certificateMetadataSchema = z.object({
  issuer: certificateIssuerSchema,
  unit: certificateUnitSchema,
  name: z.string(),
  events: z.array(certificateEventSchema),
  statuses: z.array(CertificateStatusEnum),
  id: z.string(),
  issued: z.string().datetime(),
  type: certificateTypeSchema,
  summary: certificateSummarySchema,
})

export const certificateContentSchema = z.object({ heading: z.string(), body: z.string() })

export const certificateSchema = z.object({
  metadata: certificateMetadataSchema,
  content: z.array(certificateContentSchema),
})

export type CertificateContent = z.infer<typeof certificateContentSchema>
export type CertificateEvent = z.infer<typeof certificateEventSchema>
export type CertificateIssuer = z.infer<typeof certificateIssuerSchema>
export type CertificateMetadata = z.infer<typeof certificateMetadataSchema>
export type CertificateStatus = z.infer<typeof CertificateStatusEnum>
export type CertificateSummary = z.infer<typeof certificateSummarySchema>
export type CertificateUnit = z.infer<typeof certificateUnitSchema>
export type Certificate = z.infer<typeof certificateSchema>
