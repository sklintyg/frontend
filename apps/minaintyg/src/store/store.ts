import { configureStore } from '@reduxjs/toolkit'
import { api } from './api'
import { errorMiddleware } from './middleware/error.middleware'
import { reducer } from './reducer'
import { idpMiddleware } from './slice/idpMiddleware'

export const store = configureStore({
  reducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat([api.middleware, idpMiddleware, errorMiddleware]),
})

export type AppDispatch = typeof store.dispatch
