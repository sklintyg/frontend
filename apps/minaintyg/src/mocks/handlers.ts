/* eslint-disable import/no-extraneous-dependencies */
import { fakeCertificateLabel, faker, fakerFromSchema } from '@frontend/fake'
import { rest } from 'msw'
import { certificateListItemSchema } from '../schema/certificateList.schema'
import { userSchema } from '../schema/user.schema'

export const handlers = [
  rest.get('/api/user', (_, res, ctx) =>
    res(
      ctx.status(200),
      ctx.json(
        fakerFromSchema(userSchema, {
          stringMap: {
            personName: faker.name.fullName,
          },
        })()
      )
    )
  ),
  rest.post('/api/certificates', (_, res, ctx) =>
    res(
      ctx.status(200),
      ctx.json({
        content: Array.from(
          { length: 5 },
          fakerFromSchema(certificateListItemSchema, {
            stringMap: {
              title: fakeCertificateLabel,
              certificateId: faker.datatype.uuid,
              timestamp: () => faker.date.between('2021-01-01T00:00:00.000Z', new Date().toISOString()).toISOString(),
            },
          })
        ),
      })
    )
  ),
]
