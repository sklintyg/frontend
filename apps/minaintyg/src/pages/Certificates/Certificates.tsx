import { IDSSpinner } from '@frontend/ids-react-ts'
import { useState } from 'react'
import { SortDirection } from 'react-stately'
import { PageHeading } from '../../components/PageHeading/PageHeading'
import { useGetCertificateQuery } from '../../store/api'
import { CertificateList } from './components/CertificateList'
import { CertificateListOrder } from './components/CertificateListOrder/CertificateListOrder'

export function Certificates() {
  const [order, setOrder] = useState<SortDirection>('descending')
  const { isLoading, data } = useGetCertificateQuery()

  return (
    <>
      <PageHeading heading="Intyg">
        Här listas dina läkarintyg som vården utfärdat digitalt. Hittar du inte ditt intyg, vänd dig till din mottagning. Du kan skicka
        intyg digitalt till Försäkringskassan och Transportstyrelsen. Läkarintyg om arbetsförmåga kan inte skickas digitalt till din
        arbetsgivare.
      </PageHeading>
      <CertificateListOrder setOrder={setOrder} order={order} />
      {isLoading && (
        <div data-testid="certificate-list-spinner">
          <IDSSpinner />
        </div>
      )}
      {data && <CertificateList certificates={data.content} order={order} />}
    </>
  )
}
