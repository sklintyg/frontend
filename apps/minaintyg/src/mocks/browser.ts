/* eslint-disable import/no-extraneous-dependencies */
import { faker } from 'fake'
import { setupWorker } from 'msw'
import { handlers } from './handlers'

faker.setLocale('sv')

export const worker = setupWorker(...handlers)
