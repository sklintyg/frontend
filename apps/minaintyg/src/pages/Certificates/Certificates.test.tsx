import { fakerFromSchema } from '@frontend/fake'
import { render, screen, waitFor } from '@testing-library/react'
import { rest } from 'msw'
import { Provider } from 'react-redux'
import { createMemoryRouter, createRoutesFromChildren, Route, RouterProvider } from 'react-router-dom'
import { server } from '../../mocks/server'
import { certificateFilterOptionsSchema } from '../../schema/certificateListFilter.schema'
import { store } from '../../store/store'
import { Certificates } from './Certificates'

function renderComponent() {
  render(
    <Provider store={store}>
      <RouterProvider
        router={createMemoryRouter(createRoutesFromChildren([<Route key="root" path="/" element={<Certificates />} />]), {
          initialEntries: ['/'],
        })}
      />
    </Provider>
  )
}

it('Should have correct heading', () => {
  renderComponent()
  expect(screen.getByRole('heading', { level: 1 })).toMatchSnapshot()
})

it('Should render alert message when list is empty', async () => {
  renderComponent()
  server.use(rest.post('/api/certificate', (_, res, ctx) => res(ctx.status(200), ctx.json({ content: [] }))))
  server.use(
    rest.get('/api/filters', (_, res, ctx) => res(ctx.status(200), ctx.json(fakerFromSchema(certificateFilterOptionsSchema)({ total: 0 }))))
  )
  expect(await screen.findByRole('alert')).toMatchSnapshot()
})

it('Should have correct paragraph', () => {
  renderComponent()
  expect(screen.getByText(/här listas dina läkarintyg/i)).toMatchSnapshot()
})

it('Should render list of certificates', async () => {
  renderComponent()
  expect(screen.getByTestId('certificate-list-spinner')).toBeInTheDocument()
  await waitFor(() => expect(screen.queryByTestId('certificate-list-spinner')).not.toBeInTheDocument())
})
