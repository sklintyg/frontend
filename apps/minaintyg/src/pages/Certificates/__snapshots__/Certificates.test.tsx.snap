// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should have correct heading 1`] = `
<h1
  class="ids-heading-1"
>
  Intyg
</h1>
`;

exports[`Should have correct paragraph 1`] = `
<p
  class="ids-preamble mb-2 line-clamp-5 md:line-clamp-none"
  data-clamped="true"
>
  Här listas dina läkarintyg som vården utfärdat digitalt. Hittar du inte ditt intyg, vänd dig till din mottagning. Du kan skicka intyg digitalt till Försäkringskassan och Transportstyrelsen. Läkarintyg om arbetsförmåga kan inte skickas digitalt till din arbetsgivare.
</p>
`;

exports[`Should render alert message when list is empty 1`] = `
<ids-alert
  role="alert"
>
  <p>
    I Intyg visas endast intyg som utfärdats digitalt till dig. Intyg har ingen ombudsfunktion och du kan inte ladda upp egna intyg.
  </p>
  <p>
    Om du saknar ett intyg, vänd dig till din mottagning.
  </p>
</ids-alert>
`;
